### DISCUSSION TITLE
Major features recently finished and on the horizon

### DISCUSSION THREAD

**Discussion by jryans** [koka-lang/koka#573] [General]
First off, thanks to everyone involved in the Koka project! It seems like quite interesting work. 😄 

As someone just starting to dig into Koka, I am trying to understand the current focus of the project and what's planned for the future. I can see that major version changed from 2 to 3 this year, but it's not clear to me what that's meant to indicate. The README emphasises that "Koka v3 is a research language that is currently under development and not quite ready for production use"... Is that supposed to imply Koka v2 is more stable?

Are there major features that distinguish Koka 2 and 3? I saw the [what's new document](https://github.com/koka-lang/koka/blob/dev/whatsnew.md), but I did not see anything that made it obvious why the major version changed...

I'm also curious about any feature planned for the future. I saw the list of tasks in the README. Perhaps there are also features the maintainers are focused on as well?

Apologies if I've missed something that covers these topics... I'm sure some of the answers are obvious to those who've been around for a while.

**Comment by TimWhiting** [comment]
First of all, yes Koka is very much still under development. `v2` is no more stable than `v3`. 

You should definitely use `v3`, it includes major new features:
- Implicit parameters
- Locally qualified namespaces
- A huge overhaul of the std core library - splitting into smaller modules
- Huge improvements to the compilation process (massively parallel building)
- Language server support

I think the main reason we needed a version bump was the locally qualified namespaces which changed how we organized the entire standard library, and changed naming of a few core functions.

As far as stability of v3 and usage in practice: I and a few other people from the community have gathered on Discord [invite link](https://discord.com/invite/CEkHHnTqEv) and a github community [organization](https://github.com/koka-community) to discuss and ask general "koka in the real-world" questions, and contribute common code that we can all benefit from. Discord is probably a better place for questions that need quick answers, or questions related to the community repos. Github discussions / issues on this repo are better places for talking about language development and type system issues / feedback (though some type system questions can be answered by the community).

As far as new upcoming features: I am not entirely qualified to speak to this, @daanx is the one to answer that. However, I believe one of the biggest priorities is to get a good async library with better network and file I/O support, and timers, with multiprocessing support (likely via libUV).
